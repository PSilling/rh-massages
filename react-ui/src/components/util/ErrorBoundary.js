// react imports
import React, { Component } from "react";
import PropTypes from "prop-types";

// module imports
import { Button } from "reactstrap";

// util imports
import _t from "../../util/Translations";

/**
 * ErrorBoundary class. Catches any unexpected errors and renders a reload page notification.
 */
class ErrorBoundary extends Component {
  state = { error: null, errorInfo: null };

  componentDidCatch(error, errorInfo) {
    console.error("Unexpected " + error.toString() + ": " + errorInfo.componentStack); /* eslint-disable-line */
    this.setState({ error, errorInfo });
  }

  render() {
    if (this.state.error !== null) {
      return (
        <div className="mt-3">
          <h2>
            {_t.translate("An error occured!")}
            <Button className="float-right" outline onClick={() => window.location.reload()}>
              {_t.translate("Reload page")}
            </Button>
          </h2>
          <hr />
          <details style={{ cursor: "pointer" }}>
            {`${this.state.error.toString()}: `}
            <br />
            {this.state.errorInfo.componentStack}
          </details>
        </div>
      );
    }
    return this.props.children;
  }
}

ErrorBoundary.propTypes = {
  children: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.node), PropTypes.node]).isRequired
};

export default ErrorBoundary;
