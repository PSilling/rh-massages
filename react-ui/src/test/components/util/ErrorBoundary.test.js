// react imports
import React from "react";
import { shallow } from "enzyme";

// test imports
import ErrorBoundary from "../../../components/util/ErrorBoundary";

test("renders error content correctly", () => {
  const wrapper = shallow(
    <ErrorBoundary>
      <div>test</div>
    </ErrorBoundary>
  );
  const div = wrapper.find("div");

  expect(div.props().children).toEqual("test");
  wrapper.instance().setState({ error: "error", errorInfo: { componentStack: "stack" } });
  expect(wrapper).toMatchSnapshot();
});
