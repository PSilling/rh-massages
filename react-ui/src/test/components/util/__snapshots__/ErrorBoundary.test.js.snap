// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders error content correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ErrorBoundary>
    <div>
      test
    </div>
  </ErrorBoundary>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h2>
          An error occured!
          <Button
            className="float-right"
            color="secondary"
            onClick={[Function]}
            outline={true}
            tag="button"
          >
            Reload page
          </Button>
        </h2>,
        <hr />,
        <details
          style={
            Object {
              "cursor": "pointer",
            }
          }
        >
          error: 
          <br />
          stack
        </details>,
      ],
      "className": "mt-3",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "An error occured!",
            <Button
              className="float-right"
              color="secondary"
              onClick={[Function]}
              outline={true}
              tag="button"
            >
              Reload page
            </Button>,
          ],
        },
        "ref": null,
        "rendered": Array [
          "An error occured!",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": "Reload page",
              "className": "float-right",
              "color": "secondary",
              "onClick": [Function],
              "outline": true,
              "tag": "button",
            },
            "ref": null,
            "rendered": "Reload page",
            "type": [Function],
          },
        ],
        "type": "h2",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": "hr",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "error: ",
            <br />,
            "stack",
          ],
          "style": Object {
            "cursor": "pointer",
          },
        },
        "ref": null,
        "rendered": Array [
          "error: ",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": "br",
          },
          "stack",
        ],
        "type": "details",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h2>
            An error occured!
            <Button
              className="float-right"
              color="secondary"
              onClick={[Function]}
              outline={true}
              tag="button"
            >
              Reload page
            </Button>
          </h2>,
          <hr />,
          <details
            style={
              Object {
                "cursor": "pointer",
              }
            }
          >
            error: 
            <br />
            stack
          </details>,
        ],
        "className": "mt-3",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "An error occured!",
              <Button
                className="float-right"
                color="secondary"
                onClick={[Function]}
                outline={true}
                tag="button"
              >
                Reload page
              </Button>,
            ],
          },
          "ref": null,
          "rendered": Array [
            "An error occured!",
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": "Reload page",
                "className": "float-right",
                "color": "secondary",
                "onClick": [Function],
                "outline": true,
                "tag": "button",
              },
              "ref": null,
              "rendered": "Reload page",
              "type": [Function],
            },
          ],
          "type": "h2",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": "hr",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "error: ",
              <br />,
              "stack",
            ],
            "style": Object {
              "cursor": "pointer",
            },
          },
          "ref": null,
          "rendered": Array [
            "error: ",
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": "br",
            },
            "stack",
          ],
          "type": "details",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
